chia/full_node/block_height_map.py

```diff
@@ -0,0 +1,204 @@
+        self.__height_to_hash_filename = blockchain_dir / "height-to-hash"
+        self.__ses_filename = blockchain_dir / "sub-epoch-summaries"
+
+        res = await self.db.execute(
+            "SELECT header_hash,prev_hash,height,sub_epoch_summary from block_records WHERE is_peak=1"
+        )
+        row = await res.fetchone()
+        await res.close()
+
+        if row is None:
+            return self
+
+        try:
+            async with aiofiles.open(self.__height_to_hash_filename, "rb") as f:
+                self.__height_to_hash = bytearray(await f.read())
+        except Exception:
+            # it's OK if this file doesn't exist, we can rebuild it
+            pass
+
+        try:
+            async with aiofiles.open(self.__ses_filename, "rb") as f:
+                self.__sub_epoch_summaries = {k: v for (k, v) in SesCache.from_bytes(await f.read()).content}
+        except Exception:
+            # it's OK if this file doesn't exist, we can rebuild it
+            pass
+
+        peak: bytes32 = bytes.fromhex(row[0])
+        prev_hash: bytes32 = bytes.fromhex(row[1])
+        height = row[2]
+
+        # allocate memory for height to hash map
+        # this may also truncate it, if thie file on disk had an invalid size
+        new_size = (height + 1) * 32
+        size = len(self.__height_to_hash)
+        if size > new_size:
+            del self.__height_to_hash[new_size:]
+        else:
+            self.__height_to_hash += bytearray([0] * (new_size - size))
+
+        # if the peak hash is already in the height-to-hash map, we don't need
+        # to load anything more from the DB
+        if self.get_hash(height) != peak:
+            self.__set_hash(height, peak)
+
+            if row[3] is not None:
+                self.__sub_epoch_summaries[height] = row[3]
+
+            # prepopulate the height -> hash mapping
+            await self._load_blocks_from(height, prev_hash)
+
+            await self.maybe_flush()
+
+        return self
+
+    def update_height(self, height: uint32, header_hash: bytes32, ses: Optional[SubEpochSummary]):
+        # we're only updating the last hash. If we've reorged, we already rolled
+        # back, making this the new peak
+        idx = height * 32
+        assert idx <= len(self.__height_to_hash)
+        self.__height_to_hash[idx : idx + 32] = header_hash
+        if ses is not None:
+            self.__sub_epoch_summaries[height] = bytes(ses)
+
+    async def maybe_flush(self):
+        if self.__dirty < 1000:
+            return
+
+        assert (len(self.__height_to_hash) % 32) == 0
+        map_buf = self.__height_to_hash.copy()
+
+        ses_buf = bytes(SesCache([(k, v) for (k, v) in self.__sub_epoch_summaries.items()]))
+
+        self.__dirty = 0
+
+        # TODO: use a more sophisticated way of saving these files that doesn't
+        # destroy the existing ones
+        async with aiofiles.open(self.__height_to_hash_filename, "wb+") as f:
```

Comment:
I would like to change this to use `write_file_async()` from https://github.com/Chia-Network/chia-blockchain/pull/9079